File: practice/algorithms/implementation/modified_kaprekar_numbers/ModifiedKaprekarNumbers.scala
import scala.io.Source

object ModifiedKaprekarNumbers extends App {
  val lines = Source.stdin.getLines().map(_.toInt).toList
  val lower = lines(0)
  val upper = lines(1)
  val kaprekarNums = (lower to upper).filter(isKaprekar)
  val output = if(!kaprekarNums.isEmpty) kaprekarNums.mkString(" ") else "INVALID RANGE"

  def isKaprekar(num: Int): Boolean = {
    val square = Math.pow(num, 2).toInt.toString
    if(square.length > 1) {
      val parts = square.splitAt(square.length / 2)
      val l = parts._1.toInt
      val r = parts._2.toInt
      return (l + r) == num
    } else {
      return square.toInt == num
    }
  }
}

File: practice/algorithms/implementation/modified_kaprekar_numbers/ModifiedKaprekarNumbers.java
import java.util.*;

public class ModifiedKaprekarNumbers {

    public static void main(String[] args) {
        Scanner stdin = new Scanner(System.in);
        int lower = stdin.nextInt();
        int upper = stdin.nextInt();
        List<Integer> kaprekars = new ArrayList<>();
        for(int i = lower; i <= upper; i++) {
            if(isKaprekar(i)) {
                kaprekars.add(i);
            }
        }
        if(kaprekars.isEmpty()) {
            System.out.println("INVALID RANGE");
        } else {
            for(int k : kaprekars) {
                System.out.print(k + " ");
            }
        }
        stdin.close();
    }

    private static boolean isKaprekar(int num) {
        String square = Long.valueOf((long) Math.pow(num, 2)).toString();
        if(square.length() > 1) {
            long l = Long.parseLong(square.substring(0, square.length() / 2));
            long r = Long.parseLong(square.substring(square.length() / 2));
            return (l + r) == num;
        } else {
            return Long.parseLong(square) == num;
        }
    }
}